src/Dialect/ONNX/ONNXOps/Math/Clip.cpp

```diff
@@ -41,46 +59,80 @@ LogicalResult ONNXClipOpShapeHelper::computeShape() {
 // Shape Inference
 //===----------------------------------------------------------------------===//
 
-LogicalResult ONNXClipOp::inferShapes(
-    std::function<void(Region &)> doShapeInference) {
+template <typename T>
+LogicalResult ONNXClipOpGeneralInferShapes(
```

Comment:
The pattern I have been using is when we have a custom function, I add it to the class. See this example for reference:
```C++
struct ONNXBroadcastOpShapeHelper : public ONNXOpShapeHelper {
  ONNXBroadcastOpShapeHelper(mlir::Operation *op, mlir::ValueRange operands,
      IndexExprBuilder *ieBuilder = nullptr, IndexExprScope *scope = nullptr,
      bool hasUniBroadcasting = false, bool hasNoBroadcasting = false)
      : ONNXOpShapeHelper(op, operands, ieBuilder, scope), inputsDims(),
        outputRank(0), hasUniBroadcasting(hasUniBroadcasting),
        hasNoBroadcasting(hasNoBroadcasting) {}
  virtual ~ONNXBroadcastOpShapeHelper() {}

  // Custom shape compute which takes additional parameters.
  mlir::LogicalResult customComputeShape(
      mlir::ValueRange initialOperands, DimsExpr *additionalOperand);

  // Default shape compute (every operands of the operation and no additional
  // parameters).
  mlir::LogicalResult computeShape() override {
    return customComputeShape(operands, nullptr);
  }
...
```

Here the default `computeShape` uses the specialized `customComputeShape` with default parameters, and then the ones that needs special computations, say
```C++
// Helper for ExpandOp
struct ONNXExpandOpShapeHelper : public ONNXBroadcastOpShapeHelper {
  ONNXExpandOpShapeHelper(mlir::Operation *op, mlir::ValueRange operands,
      IndexExprBuilder *ieBuilder = nullptr, IndexExprScope *scope = nullptr)
      : ONNXBroadcastOpShapeHelper(op, operands, ieBuilder, scope) {}
  virtual ~ONNXExpandOpShapeHelper() {}
  mlir::LogicalResult computeShape() final;
};
```
 
are defined and in their `computeShape()` call we can put the extra processing to generate the specialized parameters that are needed for their parent class `customComputeShape` which will compute the shape with the different input params.

Reply:
@AlexandreEichenberger Thank you very much for your comments.
Because output of onnx.ClipOp has the same shape and element type of its first input argument, it can be considered as an elementsize unary operation. I updated this PR by using the ONNXUnaryOpShapeHelper for onnx.Clip/V6/V11/V12 ops.